src/zenml/integrations/hyperai/orchestrators/hyperai_orchestrator.py

```diff
@@ -0,0 +1,336 @@
+                    }
+                }
+
+        # Convert into yaml
+        logger.info("Finalizing Docker Compose definition.")
+        compose_definition = yaml.dump(compose_definition)
+
+        # Connect to configured HyperAI instance
+        logger.info(
+            "Connecting to HyperAI instance and placing Docker Compose file."
+        )
+        paramiko_client: paramiko.SSHClient
+        if connector := self.get_connector():
+            paramiko_client = connector.connect()
+            if paramiko_client is None:
+                raise RuntimeError(
+                    "Expected to receive a `paramiko.SSHClient` object from the "
+                    "linked connector, but got `None`. This likely originates from "
+                    "a misconfigured service connector, typically caused by a wrong "
+                    "SSH key type being selected. Please check your "
+                    "`hyperai_orchestrator` configuration and make sure that the "
+                    "`ssh_key_type` of its connected service connector is set to the "
+                    "correct value."
+                )
+            elif not isinstance(paramiko_client, paramiko.SSHClient):
+                raise RuntimeError(
+                    f"Expected to receive a `paramiko.SSHClient` object from the "
+                    f"linked connector, but got type `{type(paramiko_client)}`."
+                )
+        else:
+            raise RuntimeError(
+                "You must link a HyperAI service connector to the orchestrator."
+            )
+
+        # Get container registry autologin setting
+        if self.config.container_registry_autologin:
+            logger.info(
+                "Attempting to automatically log in to container registry used by stack."
+            )
+
+            # Select stack container registry
+            container_registry = stack.container_registry
+
+            # Raise error if no container registry is found
+            if not container_registry:
+                raise RuntimeError(
+                    "Unable to find container registry in stack."
+                )
+
+            # Get container registry credentials from its config
+            container_registry_url = container_registry.config.uri
+            (
+                container_registry_username,
+                container_registry_password,
+            ) = container_registry.credentials
+
+            # Log in to container registry using --password-stdin
+            stdin, stdout, stderr = paramiko_client.exec_command(
+                f"docker login -u {container_registry_username} --password-stdin {container_registry_url} <<< {container_registry_password}"
+            )
+
+            # Log stdout
+            for line in stdout.readlines():
+                logger.info(line)
+
+        # Set up pipeline-runs directory if it doesn't exist
+        nonscheduled_directory_name = "/home/zenml/pipeline-runs"
```

Comment:
Very fair question. I'd like to keep things a bit standardized here especially because the cleanup mechanism has assumptions about where things are stored (i.e. if you configure the scheduled and nonscheduled directories in the same way, you risk losing pipeline files that are necessary per the schedule). Still, the dir may not be writable. However, we know the `username` with which the service connector logs in. I think it is rare that a user cannot write to their own home directory. Hence, I would like to propose to use `f"/home/{username}/pipeline-runs"` (or its scheduled equivalent) as it is likely writable all the time and then in line with service connector configuration.

Reply:
Sounds good to me!

Reply:
31d0a2849